File: python-package/src/enderecobr/padronizar_enderecos.py
from .padronizar_logradouros import padronizar_logradouros
from .padronizar_tipos_de_logradouro import padronizar_tipos_de_logradouro
from .padronizar_logradouros_completos import padronizar_logradouros_completos
from .padronizar_numeros import padronizar_numeros
from .padronizar_complementos import padronizar_complementos
from .padronizar_bairros import padronizar_bairros
from .padronizar_municipios import padronizar_municipios
from .padronizar_estados import padronizar_estados
from .padronizar_ceps import padronizar_ceps

import warnings
from typing import Dict, Any, List, Callable, Optional

import pandas as pd



# --------------------------------------------------------------------------- #
# Utilidades / validações                                                     #
# --------------------------------------------------------------------------- #

class EnderecoPadronizacaoError(ValueError):
    """Erro de validação de argumentos na padronização de endereços."""


def _check_choice(arg: str, nome: str, choices: List[str]) -> None:
    if arg not in choices:
        raise EnderecoPadronizacaoError(
            f'Valor inválido para "{nome}": {arg!r}. Opções válidas: {choices}.'
        )


def _checa_campos_do_endereco(
    campos: Dict[str, str],
    df: pd.DataFrame,
) -> None:
    """
    • Confere tipos; nomes únicos; campos permitidos; se colunas existem no DF.
    • Avisa (warning) se já existirem colunas "<campo>_padr" que serão sobrescritas.
    """
    if not isinstance(campos, dict) or not all(isinstance(k, str) for k in campos):
        raise EnderecoPadronizacaoError(
            "`campos_do_endereco` deve ser um dicionário str -> str."
        )

    campos_permitidos = {
        "tipo_de_logradouro",
        "logradouro",
        "numero",
        "complemento",
        "cep",
        "bairro",
        "municipio",
        "estado",
    }
    nomes = list(campos.keys())

    duplicados = {n for n in nomes if nomes.count(n) > 1}
    if duplicados:
        raise EnderecoPadronizacaoError(f"Nomes duplicados em campos_do_endereco: {duplicados}")

    não_permitidos = set(nomes) - campos_permitidos
    if não_permitidos:
        raise EnderecoPadronizacaoError(f"Campos não reconhecidos: {não_permitidos}")

    faltando = [v for v in campos.values() if v not in df.columns]
    if faltando:
        raise EnderecoPadronizacaoError(f"Colunas ausentes no dataframe: {faltando}")

    # aviso sobre sobrescrever "<campo>_padr"
    ja_existe = [f"{c}_padr" for c in nomes if f"{c}_padr" in df.columns]
    if ja_existe:
        warnings.warn(
            f"As colunas {', '.join(ja_existe)} já existem no dataframe e serão sobrescritas.",
            stacklevel=2,
        )


# --------------------------------------------------------------------------- #
# Auxiliar para combinar tipo/nome/número em um único logradouro completo     #
# --------------------------------------------------------------------------- #

def _int_padronizar_com_log_completo(
    df: pd.DataFrame,
    campos: Dict[str, str],
    campos_do_logradouro: List[str],
    checar_tipos: bool,
):
    """
    • Seleciona colunas listadas (se existirem) → chama padronizar_logradouros_completos().
    • Adiciona coluna 'logradouro_completo_padr' ao dataframe.
    """
    presentes = {c: col for c, col in campos.items() if c in campos_do_logradouro}
    if not presentes:
        return df  # nada a fazer

    # Renomear conforme esperado por padronizar_logradouros_completos
    # (tipo_de_logradouro → tipo, logradouro → nome_do_logradouro, numero → numero etc.)
    renomeia = {
        "tipo_de_logradouro": "tipo",
        "logradouro": "nome_do_logradouro",
        "numero": "numero",
    }
    args = {renomeia[k]: df[col] for k, col in presentes.items()}

    df["logradouro_completo_padr"] = padronizar_logradouros_completos(
        **args, checar_tipos=checar_tipos
    )
    return df


# --------------------------------------------------------------------------- #
# Tabela de relacionamento campo  ↔  função padronizadora                     #
# --------------------------------------------------------------------------- #

_RELACAO_CAMPOS: List[Dict[str, Any]] = [
    dict(nome_campo="tipo_de_logradouro", funcao="padronizar_tipos_de_logradouro"),
    dict(nome_campo="logradouro",         funcao="padronizar_logradouros"),
    dict(nome_campo="numero",             funcao="padronizar_numeros"),
    dict(nome_campo="complemento",        funcao="padronizar_complementos"),
    dict(nome_campo="cep",                funcao="padronizar_ceps"),
    dict(nome_campo="bairro",             funcao="padronizar_bairros"),
    dict(nome_campo="municipio",          funcao="padronizar_municipios"),
    dict(nome_campo="estado",             funcao="padronizar_estados"),
]


# --------------------------------------------------------------------------- #
# Função principal                                                            #
# --------------------------------------------------------------------------- #

def padronizar_enderecos(
    enderecos: pd.DataFrame,
    campos_do_endereco: Dict[str, str],
    *,
    formato_estados: str = "por_extenso",
    formato_numeros: str = "character",
    manter_cols_extras: bool = True,
    combinar_logradouro: bool = False,
    checar_tipos: bool = False,
) -> pd.DataFrame:
    """
    Padroniza simultaneamente múltiplos campos de endereço contidos em um DataFrame.

    Parâmetros
    ----------
    enderecos : pandas.DataFrame
        Tabela original.
    campos_do_endereco : dict
        Mapeamento 'campo lógico' → 'nome da coluna no DataFrame'.
        Exemplo:
            {
              "logradouro": "rua",
              "numero": "nro",
              "cep": "zip",
              "estado": "uf",
              ...
            }
    formato_estados : {"por_extenso", "sigla"}
        Retorno de estados.
    formato_numeros : {"character", "integer"}
        Retorno de números.
    manter_cols_extras : bool
        Se colunas não listadas devem permanecer no resultado.
    combinar_logradouro : bool
        Se deve combinar tipo + nome + número em `logradouro_completo_padr`.
    checar_tipos : bool
        (Só com `combinar_logradouro=True`). Verifica duplicidades
        tipo/nome (ex.: "RUA" + "RUA BOTAFOGO").

    Retorno
    -------
    pandas.DataFrame – cópia do input com novas colunas *_padr (padronizadas).
    """
    if not isinstance(enderecos, pd.DataFrame):
        raise EnderecoPadronizacaoError("`enderecos` precisa ser um pandas.DataFrame.")

    _check_choice(formato_estados, "formato_estados", ["por_extenso", "sigla"])
    _check_choice(formato_numeros, "formato_numeros", ["character", "integer"])
    _checa_campos_do_endereco(campos_do_endereco, enderecos)

    df = enderecos.copy()

    # Combina tipo/nome/número em logradouro completo, se solicitado
    if combinar_logradouro:
        df = _int_padronizar_com_log_completo(
            df,
            campos_do_endereco,
            campos_do_logradouro=["tipo_de_logradouro", "logradouro", "numero"],
            checar_tipos=checar_tipos,
        )

    # Mapeia nome_campo → função real + kwargs extras
    _dispatch: Dict[str, Callable] = {
        "padronizar_tipos_de_logradouro": padronizar_tipos_de_logradouro,
        "padronizar_logradouros":         padronizar_logradouros,
        "padronizar_numeros":             padronizar_numeros,
        "padronizar_complementos":        padronizar_complementos,
        "padronizar_ceps":                padronizar_ceps,
        "padronizar_bairros":             padronizar_bairros,
        "padronizar_municipios":          padronizar_municipios,
        "padronizar_estados":             padronizar_estados,
    }

    for item in _RELACAO_CAMPOS:
        campo = item["nome_campo"]
        if campo not in campos_do_endereco:
            continue  # usuário não pediu esse campo

        col_original = campos_do_endereco[campo]
        col_padron   = f"{campo}_padr"
        func         = _dispatch[item["funcao"]]

        # kwargs extras
        extras: Dict[str, Any] = {}
        if campo == "estado":
            extras["formato"] = formato_estados
        if campo == "numero":
            extras["formato"] = formato_numeros

        df[col_padron] = func(df[col_original], **extras)  # type: ignore[arg-type]

    # --------  lidar com colunas extras  -----------------------------------
    if not manter_cols_extras:
        cols_padronizadas = [f"{c}_padr" for c in campos_do_endereco]
        if combinar_logradouro:
            cols_padronizadas.append("logradouro_completo_padr")
        df = df[cols_padronizadas]
    else:
        # põe extras primeiro (estilo data.table::setcolorder)
        extras = [c for c in enderecos.columns if c not in campos_do_endereco.values()]
        df = df[extras + [c for c in df.columns if c not in extras]]

    return df

File: r-package/R/padronizar_enderecos.R
#' Padronizar endereços
#'
#' Padroniza simultaneamente os diversos campos de um endereço listados em um
#' dataframe.
#'
#' @param enderecos Um dataframe. Os endereços a serem padronizados. Cada uma de
#'   suas colunas deve corresponder a um campo do endereço (e.g. logradouro,
#'   cidade, bairro, etc).
#' @param campos_do_endereco Um vetor nomeado de caracteres. A correspondência
#'   entre os campos a serem padronizados (nomes do vetor) e as colunas que os
#'   representam no dataframe (valores em si). A função
#'   `correspondencia_campos()` facilita a criação deste vetor, fazendo também
#'   algumas verificações do conteúdo imputado. Argumentos dessa função com
#'   valor `NULL` são ignorados, e ao menos um valor diferente de nulo deve ser
#'   fornecido. Caso deseje criar o vetor manualmente, note que seus nomes devem
#'   ser os mesmos nomes dos parâmetros da função `correspondencia_campos()`.
#' @param formato_estados Uma string. Como o estado padronizado deve ser
#'   formatado. Por padrão, `"por_extenso"`, fazendo com que a função retorne o
#'   nome dos estados por extenso. Se `"sigla"`, a função retorna a sigla dos
#'   estados.
#' @param formato_numeros Uma string. Como o número padronizado deve ser
#'   formatado. Por padrão, `"character"`, fazendo com que a função retorne o
#'   número como caractere. Se `"integer"`, a função retorna o número como
#'   inteiro.
#' @param manter_cols_extras Um logical. Se colunas não especificadas em
#'   `campos_do_endereco` devem ser mantidas ou não (por exemplo, uma coluna de
#'   id do conjunto de dados sendo padronizado). Por padrão, `TRUE`.
#' @param combinar_logradouro Um logical. Se os campos que descrevem o
#'   logradouro (tipo, nome e número, por exemplo) devem ser combinados em um
#'   único campo de logradouro completo. Nesse caso, o parâmetro `logradouro` da
#'   `correspondencia_campos()` deve ser interpretado como o nome do logradouro.
#'   Por padrão, `FALSE`.
#' @param checar_tipos Um logical. Apenas tem efeito quando
#'   `combinar_logradouro` é `TRUE`. Se a ocorrência de duplicatas entre os
#'   tipos e nomes dos logradouros deve ser verificada ao combiná-los (por
#'   exemplo, quando o tipo é descrito como "RUA" e o nome é descrito como "RUA
#'   BOTAFOGO"). Por padrão, `FALSE`.
#'
#' @return Um dataframe com colunas adicionais, representando os campos de
#'   endereço padronizados.
#'
#' @examples
#' enderecos <- data.frame(
#'   id = 1,
#'   logradouro = "r ns sra da piedade",
#'   nroLogradouro = 20,
#'   complemento = "qd 20",
#'   cep = 25220020,
#'   bairro = "jd botanico",
#'   codmun_dom = 3304557,
#'   uf_dom = "rj"
#' )
#'
#' campos <- correspondencia_campos(
#'   logradouro = "logradouro",
#'   numero = "nroLogradouro",
#'   complemento = "complemento",
#'   cep = "cep",
#'   bairro = "bairro",
#'   municipio = "codmun_dom",
#'   estado = "uf_dom"
#' )
#'
#' padronizar_enderecos(enderecos, campos)
#'
#' padronizar_enderecos(enderecos, campos, manter_cols_extras = FALSE)
#'
#' padronizar_enderecos(enderecos, campos, combinar_logradouro = TRUE)
#'
#' ends_tipo_duplicado <- data.frame(tipo = "r", nome = "r ns sra da piedade")
#'
#' padronizar_enderecos(
#'   ends_tipo_duplicado,
#'   campos_do_endereco = correspondencia_campos(
#'     tipo_de_logradouro = "tipo",
#'     logradouro = "nome"
#'   ),
#'   combinar_logradouro = TRUE,
#'   checar_tipos = TRUE
#' )
#'
#' @export
padronizar_enderecos <- function(
  enderecos,
  campos_do_endereco = correspondencia_campos(),
  formato_estados = "por_extenso",
  formato_numeros = "character",
  manter_cols_extras = TRUE,
  combinar_logradouro = FALSE,
  checar_tipos = FALSE
) {
  checkmate::assert_data_frame(enderecos)
  checkmate::assert(
    checkmate::check_string(formato_estados),
    checkmate::check_names(
      formato_estados,
      subset.of = c("por_extenso", "sigla")
    ),
    combine = "and"
  )
  checkmate::assert(
    checkmate::check_string(formato_numeros),
    checkmate::check_names(
      formato_numeros,
      subset.of = c("character", "integer")
    ),
    combine = "and"
  )
  checkmate::assert_logical(manter_cols_extras, any.missing = FALSE, len = 1)
  checkmate::assert_logical(combinar_logradouro, any.missing = FALSE, len = 1)
  checkmate::assert_logical(checar_tipos, any.missing = FALSE, len = 1)
  checa_campos_do_endereco(campos_do_endereco, enderecos)

  enderecos_padrao <- data.table::as.data.table(enderecos)

  relacao_campos <- tibble::tribble(
    ~nome_campo,          ~nome_formatado,       ~funcao,                        ~args_extra,
    "tipo_de_logradouro", "tipos de logradouro", padronizar_tipos_de_logradouro, NULL,
    "logradouro",         "logradouros",         padronizar_logradouros,         NULL,
    "numero",             "n\u00fameros",        padronizar_numeros,             list(formato = formato_numeros),
    "complemento",        "complementos",        padronizar_complementos,        NULL,
    "cep",                "CEPs",                padronizar_ceps,                NULL,
    "bairro",             "bairros",             padronizar_bairros,             NULL,
    "municipio",          "munic\u00edpios",     padronizar_municipios,          NULL,
    "estado",             "estados",             padronizar_estados,             list(formato = formato_estados)
  )

  if (combinar_logradouro) {
    campos_do_logradouro <- c("tipo_de_logradouro", "logradouro", "numero")

    enderecos_padrao <- int_padronizar_ends_com_log_compl(
      enderecos_padrao,
      campos_do_endereco,
      campos_do_logradouro,
      checar_tipos
    )

    relacao_campos <- subset(
      relacao_campos,
      ! nome_campo %in% campos_do_logradouro
    )
  }

  invisible(
    mapply(
      relacao_campos$nome_campo,
      relacao_campos$nome_formatado,
      relacao_campos$funcao,
      relacao_campos$args_extra,
      FUN = function(nome_campo, nome_formatado, funcao, args_extra) {
        if (nome_campo %in% names(campos_do_endereco)) {
          col_orig <- campos_do_endereco[nome_campo]
          col_padr <- paste0(nome_campo, "_padr")

          prog <- mensagem_progresso_endbr(
            paste0("Padronizando ", nome_formatado, "...")
          )

          enderecos_padrao[
            ,
            c(col_padr) := do.call(
              funcao,
              args = append(list(enderecos[[col_orig]]), args_extra)
            )
          ]

          cli::cli_progress_done(id = prog)
        }
      }
    )
  )

  campos_extras <- setdiff(names(enderecos), campos_do_endereco)

  if (!manter_cols_extras) {
    enderecos_padrao[, (campos_extras) := NULL]
  } else {
    data.table::setcolorder(enderecos_padrao, campos_extras)
  }

  return(enderecos_padrao[])
}

int_padronizar_ends_com_log_compl <- function(enderecos_padrao,
                                              campos_do_endereco,
                                              campos_do_logradouro,
                                              checar_tipos) {
  campos_do_log_listados <- campos_do_endereco[
    which(names(campos_do_endereco) %in% campos_do_logradouro)
  ]

  if (length(campos_do_log_listados) > 0) {
    names(campos_do_log_listados) <- sub(
      "^logradouro$",
      "nome_do_logradouro",
      names(campos_do_log_listados)
    )

    enderecos_padrao <- padronizar_logradouros_completos(
      enderecos_padrao,
      campos_do_logradouro = campos_do_log_listados,
      checar_tipos = checar_tipos
    )
  }

  return(enderecos_padrao)
}

checa_campos_do_endereco <- function(campos_do_endereco, enderecos) {
  col <- checkmate::makeAssertCollection()
  checkmate::assert_character(
    campos_do_endereco,
    any.missing = FALSE,
    add = col
  )
  checkmate::assert_names(
    names(campos_do_endereco),
    type = "unique",
    subset.of = c(
      "tipo_de_logradouro",
      "logradouro",
      "numero",
      "complemento",
      "cep",
      "bairro",
      "municipio",
      "estado"
    ),
    add = col
  )
  checkmate::assert_names(
    campos_do_endereco,
    subset.of = names(enderecos),
    add = col
  )
  checkmate::reportAssertions(col)

  # a função retorna os valores padronizados em colunas de nome "<campo>_padr".
  # a seguir, checamos se colunas com esse nome existem e lançamos um warning
  # caso positivo

  campos_padr <- paste0(names(campos_do_endereco), "_padr")

  campos_padr_existentes <- campos_padr[campos_padr %in% names(enderecos)]

  if (length(campos_padr_existentes) > 0) {
    warning_coluna_existente(campos_padr_existentes)
  }

  return(invisible(TRUE))
}

warning_coluna_existente <- function(campos_padr_existentes) {
  lista_campos <- cli::cli_vec(
    campos_padr_existentes,
    list("vec-last" = " e ", "vec-sep2" = " e ")
  )

  warning_endbr(
    c(
      paste0(
        "A{?s} seguinte{?s} coluna{?s} fo{?i/ram} encontrada{?s} no input e ",
        "ser{?\u00e1/\u00e3o} sobrescrita{?s} no output: ",
        "{.var {lista_campos}}."
      )
    ),
    call = rlang::caller_env(n = 2),
    .envir = environment()
  )
}
